src/main/scala/is/hail/expr/Type.scala

```diff
@@ -824,12 +839,16 @@ final case class TArray(elementType: Type) extends TIterable {
   }
 
   def typeCheck(a: Any): Boolean = a == null || (a.isInstanceOf[IndexedSeq[_]] &&
-    a.asInstanceOf[IndexedSeq[_]].forall(elementType.typeCheck))
+    a.asInstanceOf[IndexedSeq[_]].forall(elementType.typeCheck) &&
+    (!elementsRequired || a.asInstanceOf[IndexedSeq[_]].forall(i => {assert(i != null); true})))
 
   override def str(a: Annotation): String = JsonMethods.compact(toJSON(a))
 
   override def genNonmissingValue: Gen[Annotation] =
-    Gen.buildableOf[Array, Annotation](elementType.genValue).map(x => x: IndexedSeq[Annotation])
+    if (elementsRequired)
+      Gen.buildableOf[Array, Annotation](elementType.genNonmissingValue).map(x => x: IndexedSeq[Annotation])
+    else
+      Gen.buildableOf[Array, Annotation](elementType.genValue).map(x => x: IndexedSeq[Annotation])
```

Comment:
I think I prefer modifying `TContianer`:

```scala
class TContainer {
  // ...
  def genElement: Gen[Annotation] =
    if (elementsRequired) elementType.genNonMissingValue else elementType.genValue
}
// ...
class TArray {
  // ...
  override def genNonmissingValue: Gen[Array[Annotation]] =
    Gen.buildableOf[Array, Annotation](genElement).map(x => x: IndexedSeq[Annotation])
}
```

Reply:
that makes sense. since elementsRequired is defined on TContainer and all.

Reply:
Oh this might look a bit ugly with TDict...

I'd like to see how it shakes out.

At the least I think a separate `genElement` method on each subclass is more readable than what we have now.